File: saltedge-sca-service-spring/saltedge-sca-service-sdk/src/test/java/com/saltedge/sca/sdk/interceptors/AuthorizationErrorsIntegrationTests.java
/*
 * This file is part of the Salt Edge Authenticator distribution
 * (https://github.com/saltedge/sca-identity-service-example).
 * Copyright (c) 2020 Salt Edge Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 or later.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 * For the additional permissions granted for Salt Edge Authenticator
 * under Section 7 of the GNU General Public License see THIRD_PARTY_NOTICES.md
 */
package com.saltedge.sca.sdk.interceptors;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.saltedge.sca.sdk.TestTools;
import com.saltedge.sca.sdk.models.api.ErrorResponse;
import com.saltedge.sca.sdk.models.api.requests.UpdateAuthorizationRequest;
import com.saltedge.sca.sdk.models.api.responces.AuthorizationsResponse;
import com.saltedge.sca.sdk.models.api.responces.UpdateAuthorizationResponse;
import com.saltedge.sca.sdk.models.persistent.AuthorizationEntity;
import com.saltedge.sca.sdk.models.persistent.AuthorizationsRepository;
import com.saltedge.sca.sdk.models.persistent.ClientConnectionEntity;
import com.saltedge.sca.sdk.models.persistent.ClientConnectionsRepository;
import com.saltedge.sca.sdk.provider.ServiceProvider;
import com.saltedge.sca.sdk.tools.DateTools;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.web.server.LocalServerPort;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.util.LinkedMultiValueMap;

import javax.naming.NamingException;
import java.io.FileNotFoundException;
import java.time.LocalDateTime;
import java.util.ArrayList;

import static com.saltedge.sca.sdk.ScaSdkConstants.*;
import static com.saltedge.sca.sdk.controllers.AuthorizationsController.AUTHORIZATIONS_REQUEST_PATH;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.BDDMockito.given;

@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class AuthorizationErrorsIntegrationTests {
	@LocalServerPort
	Integer randomServerPort = 0;
	@MockBean
	private ServiceProvider serviceProvider;
	@MockBean
	private ClientConnectionsRepository connectionsRepository;
	@MockBean
	private AuthorizationsRepository authorizationsRepository;

	private TestRestTemplate testRestTemplate = new TestRestTemplate();
	private ClientConnectionEntity testConnection;
	private AuthorizationEntity testAuthorization;
	
	public AuthorizationErrorsIntegrationTests() throws Exception {
		testConnection = new ClientConnectionEntity();
		try {
			testConnection.setPublicKey(TestTools.getRsaPublicKeyPemString());
		} catch (FileNotFoundException | NamingException e) {
			e.printStackTrace();
		}
		testConnection.setUserId(String.valueOf(1));
		
		testAuthorization = new AuthorizationEntity();
		testAuthorization.setId(1L);
		testAuthorization.setUserId(String.valueOf(1));
	}

	@Test
	public void getAuthorizationsTest_returnError_whenNoAccessToken() throws Exception {
		String requestUrl = getBaseUrl() + AUTHORIZATIONS_REQUEST_PATH;
		String expiresAt = String.valueOf(DateTools.nowUtcSeconds() + 60);
		String signature = TestTools.createSignature(
				HttpMethod.POST.toString(),
				requestUrl,
				expiresAt,
				"",
				TestTools.getRsaPrivateKey()
		);
		LinkedMultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add(HEADER_KEY_EXPIRES_AT, expiresAt);
		headers.add(HEADER_KEY_SIGNATURE, signature);
		ResponseEntity<ErrorResponse> response = testRestTemplate.exchange(requestUrl, HttpMethod.GET, new HttpEntity<>(headers), ErrorResponse.class);

		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
		assertThat(response.getBody()).isEqualTo(new ErrorResponse("AccessTokenMissing", "Access Token is missing."));
	}

	@Test
	public void getAuthorizationsTest_returnError_whenNoExpiresAt() {
		String requestUrl = getBaseUrl() + AUTHORIZATIONS_REQUEST_PATH;
		LinkedMultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken");

		ResponseEntity<ErrorResponse> response = testRestTemplate.exchange(requestUrl, HttpMethod.GET, new HttpEntity<>(headers), ErrorResponse.class);

		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
		assertThat(response.getBody()).isEqualTo(new ErrorResponse("SignatureExpired", "Expired Signature."));
	}

	@Test
	public void getAuthorizationsTest_returnError_whenExpiresAtIsBeforeNow() {
		String requestUrl = getBaseUrl() + AUTHORIZATIONS_REQUEST_PATH;
		LinkedMultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken");
		headers.add(HEADER_KEY_EXPIRES_AT, "0");

		ResponseEntity<ErrorResponse> response = testRestTemplate.exchange(requestUrl, HttpMethod.GET, new HttpEntity<String>(headers), ErrorResponse.class);

		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
		assertThat(response.getBody()).isEqualTo(new ErrorResponse("SignatureExpired", "Expired Signature."));
	}

	@Test
	public void getAuthorizationsTest_returnError_whenNoSignature() {
		String requestUrl = getBaseUrl() + AUTHORIZATIONS_REQUEST_PATH;
		LinkedMultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken");
		headers.add(HEADER_KEY_EXPIRES_AT, String.valueOf(DateTools.nowUtcSeconds() + 60));

		ResponseEntity<ErrorResponse> response = testRestTemplate.exchange(requestUrl, HttpMethod.GET, new HttpEntity<>(headers), ErrorResponse.class);

		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
		assertThat(response.getBody()).isEqualTo(new ErrorResponse("SignatureMissing", "Signature is missing."));
	}

	@Test
	public void getAuthorizationsTest_returnError_whenNoConnectionByAccessToken() {
		String requestUrl = getBaseUrl() + AUTHORIZATIONS_REQUEST_PATH;
		given(connectionsRepository.findByAccessTokenAndRevokedFalse("accessToken")).willReturn(null);
		LinkedMultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken");
		headers.add(HEADER_KEY_EXPIRES_AT, String.valueOf(DateTools.nowUtcSeconds() + 60));
		headers.add(HEADER_KEY_SIGNATURE, "signature");

		ResponseEntity<ErrorResponse> response = testRestTemplate.exchange(requestUrl, HttpMethod.GET, new HttpEntity<>(headers), ErrorResponse.class);

		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.UNAUTHORIZED);
		assertThat(response.getBody()).isEqualTo(new ErrorResponse("ConnectionNotFound", "Authenticator Connection Not Found."));
	}

	@Test
	public void getAuthorizationsTest_returnError_whenSignatureIsInvalid() throws Exception {
		String requestUrl = getBaseUrl() + AUTHORIZATIONS_REQUEST_PATH;
		String expiresAt = String.valueOf(DateTools.nowUtcSeconds() + 60);
		String signature = TestTools.createSignature(
				HttpMethod.POST.toString(),
				requestUrl,
				expiresAt,
				"",
				TestTools.getRsaPrivateKey()
		);
		given(connectionsRepository.findByAccessTokenAndRevokedFalse("accessToken")).willReturn(testConnection);
		LinkedMultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken");
		headers.add(HEADER_KEY_EXPIRES_AT, expiresAt);
		headers.add(HEADER_KEY_SIGNATURE, signature);

		ResponseEntity<ErrorResponse> response = testRestTemplate.exchange(requestUrl, HttpMethod.GET, new HttpEntity<>(headers), ErrorResponse.class);

		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
		assertThat(response.getBody()).isEqualTo(new ErrorResponse("InvalidSignature", "Invalid Signature."));
	}

	@Test
	public void getAuthorizationsTest_returnSuccess_onGetAuthorizations() throws Exception {
		String requestUrl = getBaseUrl() + AUTHORIZATIONS_REQUEST_PATH;
		String expiresAt = String.valueOf(DateTools.nowUtcSeconds() + 60);
		String signature = TestTools.createSignature(
				HttpMethod.GET.toString(),
				requestUrl,
				expiresAt,
				"",
				TestTools.getRsaPrivateKey()
		);

		ArgumentCaptor<String> userCaptor = ArgumentCaptor.forClass(String.class);
		given(connectionsRepository.findByAccessTokenAndRevokedFalse("accessToken")).willReturn(testConnection);
		given(authorizationsRepository.findByUserIdAndExpiresAtGreaterThanAndConfirmedIsNull(
				userCaptor.capture(),
				any(LocalDateTime.class))
		).willReturn(new ArrayList<>());

		LinkedMultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken");
		headers.add(HEADER_KEY_EXPIRES_AT, expiresAt);
		headers.add(HEADER_KEY_SIGNATURE, signature);

		ResponseEntity<AuthorizationsResponse> response = testRestTemplate.exchange(requestUrl, HttpMethod.GET, new HttpEntity<>(headers), AuthorizationsResponse.class);

		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
		assertThat(response.getBody().data).isEmpty();
	}

	@Test
	public void updateAuthorizationTest_returnSuccess_onPutAuthorization() throws Exception {
		String requestUrl = getBaseUrl() + AUTHORIZATIONS_REQUEST_PATH + "/1";
		String expiresAt = String.valueOf((DateTools.nowUtcSeconds() + 60));

		UpdateAuthorizationRequest body = new UpdateAuthorizationRequest(true, "1234567890");

		String rawBody = new ObjectMapper().writeValueAsString(body);
		String signature = TestTools.createSignature(
				HttpMethod.PUT.toString(),
				requestUrl,
				expiresAt,
				rawBody,
				TestTools.getRsaPrivateKey()
		);
		LinkedMultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken");
		headers.add(HEADER_KEY_EXPIRES_AT, expiresAt);
		headers.add(HEADER_KEY_SIGNATURE, signature);

		ArgumentCaptor<String> userCaptor = ArgumentCaptor.forClass(String.class);
		given(connectionsRepository.findByAccessTokenAndRevokedFalse("accessToken")).willReturn(testConnection);
		given(authorizationsRepository.findFirstByIdAndUserIdAndExpiresAtGreaterThanAndConfirmedIsNull(
				eq(1L),
				userCaptor.capture(),
				any(LocalDateTime.class))
		).willReturn(testAuthorization);

		ResponseEntity<UpdateAuthorizationResponse> response = testRestTemplate.exchange(requestUrl, HttpMethod.PUT, new HttpEntity<>(body, headers), UpdateAuthorizationResponse.class);

		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
		assertThat(response.getBody().data.success).isFalse();
	}

	private String getBaseUrl() {
		return "http://localhost:" + randomServerPort;
	}
}

File: spring-example/src/test/kotlin/com/saltedge/authenticator/identity/controller/AuthorizationErrorsIntegrationTests.kt
/*
 * This file is part of the Salt Edge Authenticator distribution
 * (https://github.com/saltedge/sca-identity-service-example).
 * Copyright (c) 2019 Salt Edge Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 or later.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 * For the additional permissions granted for Salt Edge Authenticator
 * under Section 7 of the GNU General Public License see THIRD_PARTY_NOTICES.md
 */
package com.saltedge.authenticator.identity.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.saltedge.authenticator.identity.HEADER_KEY_ACCESS_TOKEN
import com.saltedge.authenticator.identity.HEADER_KEY_EXPIRES_AT
import com.saltedge.authenticator.identity.HEADER_KEY_SIGNATURE
import com.saltedge.authenticator.identity.TestTools
import com.saltedge.authenticator.identity.controller.api.AUTHORIZATIONS_REQUEST_PATH
import com.saltedge.authenticator.identity.model.*
import com.saltedge.authenticator.identity.model.mapping.*
import com.saltedge.authenticator.identity.tools.nowUtcSeconds
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.anyLong
import org.mockito.ArgumentMatchers.eq
import org.mockito.BDDMockito.given
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.http.HttpEntity
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.util.LinkedMultiValueMap
import java.util.*

@RunWith(SpringRunner::class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
class AuthorizationErrorsIntegrationTests {
	@LocalServerPort
	var randomServerPort: Int = 0
	val baseUrl: String
		get() = "http://localhost:$randomServerPort"
	@MockBean
	private val connectionsRepository: ConnectionsRepository? = null
	@MockBean
	private val authorizationsRepository: AuthorizationsRepository? = null
	private val testRestTemplate: TestRestTemplate
		get() = TestRestTemplate()
	private val testConnection = Connection().apply {
		publicKey = TestTools.rsaPublicKeyString
		user = User()
	}
	private val testAuthorization = Authorization().apply {
		id = 1
		user = User()
	}

	@Test
	fun getAuthorizationsTest_returnError_whenNoAccessToken() {
		val headers = LinkedMultiValueMap<String, String>()
		val response = testRestTemplate.exchange("$baseUrl$AUTHORIZATIONS_REQUEST_PATH", HttpMethod.GET, HttpEntity<Any>(headers), ErrorResponse::class.java)

		assertThat(response.statusCode).isEqualTo(HttpStatus.UNAUTHORIZED)
		assertThat(response.body).isEqualTo(ErrorResponse("AccessTokenNotFound", "AccessTokenNotFound"))
	}

	@Test
	fun getAuthorizationsTest_returnError_whenNoExpiresAt() {
		val headers = LinkedMultiValueMap<String, String>()
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken")

		val response = testRestTemplate.exchange("$baseUrl$AUTHORIZATIONS_REQUEST_PATH", HttpMethod.GET, HttpEntity<Any>(headers), ErrorResponse::class.java)

		assertThat(response.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
		assertThat(response.body).isEqualTo(ErrorResponse("SignatureExpired", "SignatureExpired"))
	}

	@Test
	fun getAuthorizationsTest_returnError_whenExpiresAtIsBeforeNow() {
		val headers = LinkedMultiValueMap<String, String>()
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken")
		headers.add(HEADER_KEY_EXPIRES_AT, "0")

		val response = testRestTemplate.exchange("$baseUrl$AUTHORIZATIONS_REQUEST_PATH", HttpMethod.GET, HttpEntity<Any>(headers), ErrorResponse::class.java)

		assertThat(response.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
		assertThat(response.body).isEqualTo(ErrorResponse("SignatureExpired", "SignatureExpired"))
	}

	@Test
	fun getAuthorizationsTest_returnError_whenNoSignature() {
		val headers = LinkedMultiValueMap<String, String>()
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken")
		headers.add(HEADER_KEY_EXPIRES_AT, (nowUtcSeconds + 60).toString())

		val response = testRestTemplate.exchange("$baseUrl$AUTHORIZATIONS_REQUEST_PATH", HttpMethod.GET, HttpEntity<Any>(headers), ErrorResponse::class.java)

		assertThat(response.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
		assertThat(response.body).isEqualTo(ErrorResponse("SignatureNotFound", "SignatureNotFound"))
	}

	@Test
	fun getAuthorizationsTest_returnError_whenNoConnectionByAccessToken() {
		given(connectionsRepository!!.findByAccessTokenAndRevokedFalse("accessToken")).willReturn(null)
		val headers = LinkedMultiValueMap<String, String>()
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken")
		headers.add(HEADER_KEY_EXPIRES_AT, (nowUtcSeconds + 60).toString())
		headers.add(HEADER_KEY_SIGNATURE, "signature")

		val response = testRestTemplate.exchange("$baseUrl$AUTHORIZATIONS_REQUEST_PATH", HttpMethod.GET, HttpEntity<Any>(headers), ErrorResponse::class.java)

		assertThat(response.statusCode).isEqualTo(HttpStatus.UNAUTHORIZED)
		assertThat(response.body).isEqualTo(ErrorResponse("ConnectionNotFound", "ConnectionNotFound"))
	}

	@Test
	fun getAuthorizationsTest_returnError_whenSignatureIsInvalid() {
		val requestUrl = "$baseUrl$AUTHORIZATIONS_REQUEST_PATH"
		val expiresAt = (nowUtcSeconds + 60).toString()
		val signature = TestTools.createSignature(
				requestMethod = "${HttpMethod.POST}",
				requestUrl = requestUrl,
				expiresAt = expiresAt,
				requestBody = "",
				privateKey = TestTools.rsaPrivateKey
		)
		given(connectionsRepository!!.findByAccessTokenAndRevokedFalse("accessToken")).willReturn(testConnection)
		val headers = LinkedMultiValueMap<String, String>()
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken")
		headers.add(HEADER_KEY_EXPIRES_AT, expiresAt)
		headers.add(HEADER_KEY_SIGNATURE, signature)

		val response = testRestTemplate.exchange(requestUrl, HttpMethod.GET, HttpEntity<Any>(headers), ErrorResponse::class.java)

		assertThat(response.statusCode).isEqualTo(HttpStatus.BAD_REQUEST)
		assertThat(response.body).isEqualTo(ErrorResponse("InvalidSignature", "InvalidSignature"))
	}

	@Test
	fun getAuthorizationsTest_returnSuccess_onGetAuthorizations() {
		val requestUrl = "$baseUrl$AUTHORIZATIONS_REQUEST_PATH"
		val expiresAt = (nowUtcSeconds + 60).toString()
		val signature = TestTools.createSignature(
				requestMethod = "${HttpMethod.GET}",
				requestUrl = requestUrl,
				expiresAt = expiresAt,
				requestBody = "",
				privateKey = TestTools.rsaPrivateKey
		)

		val userCaptor = ArgumentCaptor.forClass(User::class.java)
		given(connectionsRepository!!.findByAccessTokenAndRevokedFalse("accessToken")).willReturn(testConnection)
		given(authorizationsRepository!!.findByUserAndExpiresAtGreaterThanAndConfirmedIsNull(
				user = userCaptor.capture(),
				currentDate = anyLong())
		).willReturn(emptyList());

		val headers = LinkedMultiValueMap<String, String>()
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken")
		headers.add(HEADER_KEY_EXPIRES_AT, expiresAt)
		headers.add(HEADER_KEY_SIGNATURE, signature)

		val response = testRestTemplate.exchange(requestUrl, HttpMethod.GET, HttpEntity<Any>(headers), AuthorizationsResponse::class.java)

		assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
		assertThat(response.body!!.data).isEqualTo(emptyList<EncryptedAuthorization>())
	}

	@Test
	fun updateAuthorizationTest_returnSuccess_onPutAuthorization() {
		val requestUrl = "$baseUrl$AUTHORIZATIONS_REQUEST_PATH/1"
		val expiresAt = (nowUtcSeconds + 60).toString()

		val body = UpdateAuthorizationRequest(UpdateAuthorizationRequestData(confirm = true, authorizationCode = "1234567890"))
		val rawBody = ObjectMapper().writeValueAsString(body)
		val signature = TestTools.createSignature(
				requestMethod = "${HttpMethod.PUT}",
				requestUrl = requestUrl,
				expiresAt = expiresAt,
				requestBody = rawBody,
				privateKey = TestTools.rsaPrivateKey
		)
		val headers = LinkedMultiValueMap<String, String>()
		headers.add(HEADER_KEY_ACCESS_TOKEN, "accessToken")
		headers.add(HEADER_KEY_EXPIRES_AT, expiresAt)
		headers.add(HEADER_KEY_SIGNATURE, signature)

		val userCaptor = ArgumentCaptor.forClass(User::class.java)
		given(connectionsRepository!!.findByAccessTokenAndRevokedFalse("accessToken")).willReturn(testConnection)
		given(authorizationsRepository!!.findByIdAndUserAndExpiresAtGreaterThanAndConfirmedIsNull(
				id = eq(1L),
				user = userCaptor.capture(),
				currentDate = anyLong())
		).willReturn(Optional.of(testAuthorization));

		val response = testRestTemplate.exchange(requestUrl, HttpMethod.PUT, HttpEntity(body, headers), UpdateAuthorizationResponse::class.java)

		assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
		assertThat(response.body!!.data.success).isFalse()
	}
}
